Upgraded aiopylgtv to v0.4.0

corneyl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Upgraded `aiopylgtv` to v0.4.0. `aiopylgtv` is a dependency of the webostv integration.
The problem with the previous used version is that it doesn't always report all sources correctly, I've verified both by using the `aiopylgtv` separately and by using the code of this PR that this problem was fixed by the newest v0.4.0.

Diff between v0.3.3 and 0.4.0: https://github.com/bendavid/aiopylgtv/compare/0.3.3...bendavid:0.4.0

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
webostv:
  - host: 192.168.1.50
    name: lgtv
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #42371
- This PR is related to issue: #41344
- Link to documentation pull request: N/A

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

corneyl committed:
Upgraded aiopylgtv to v0.4.0

frenck commented:
Thanks for bumping the dependency! However, could you please update the PR description to contain at least one (or multiple) of the following:

- A link to the release notes of this package version, and all versions in between.
- A link to the changelog of this package.
- A link to a Git(Hub) diff/compare view from the current version to the bumped version.

This allows us to review upstream changes, which is needed to decide this change is working as intended and/or if we can include it in, for example, a patch release of Home Assistant.

Thanks already! 👍

corneyl commented:
@frenck I've added the link to the diff, unfortunately there are no changelog or release notes of the new version.

Also I saw in the diff that the key file, in which the WebOS registration key is saved, changed from `KEY_FILE_NAME = ".aiopylgtv"` to `KEY_FILE_NAME = ".aiopylgtv.sqlite"`. I'm guessing this requires a re-pair with the TV after upgrade, so does that make this a breaking change?

I'll also have a look at the failing tests.

corneyl committed:
Made webostv unittests compatible with aiopylgtv 0.4.0

corneyl commented:
I've tested it with the `aiopylgtv` package separately, and indeed the keyfile format changes. It is however quite easy to convert the JSON keyfile to an Sqlite DB using the package `sqlitedict` (which is a dependency of `aiopylgtv` anyway).

How does HA handle this normally? Would it be best to add some code to migrate the keyfile from one format to another so it doesn't become a breaking change? It could be as simple as:
```python
def convert_client_keys(config_file):
    # Try to parse the file as being JSON
    with open(config_file) as json_file:
        try:
            json_conf = json.load(json_file)
        except json.JSONDecodeError:
            json_conf = None

    # If the file contained JSON, convert it to an Sqlite DB
    if json_conf:
        # Clean the json file
        os.remove(conf_file)

        # Write the data
        with SqliteDict(config_file) as conf:
            for ip, key in json_conf.items():
                conf[ip] = key
            conf.commit()
```

frenck commented:
You could try to do that on init. In general, we don't use the methods, this integration is using, anymore.

If the integration is converted to use a configuration flow, then Home Assistant provides means for migrations.

corneyl committed:
Added migration of JSON-based webostv config file to Sqlite + added unit tests

corneyl commented:
I've not enough knowledge to convert it to a configuration flow approach I'm afraid. But I've added the key-file migration on init and added some tests to test the migration. Also tested this several times locally of course.

corneyl committed:
Fixed some static code check errors in webostv init/tests

corneyl committed:
Fixed more static code check problems in webostv init/tests

corneyl committed:
Fixed even more static code check problems in webostv init/tests

corneyl committed:
Added dot to first line of comment in webostv test cleanup method

corneyl commented:
@frenck was anything else needed for this PR? Would be nice if it could be part of the next major HA release!

MartinHjelmare commented:
If we would migrate the integration to use a config flow and config entries we wouldn't need the library to store the key in the sqlite db. That's just unnecessary overhead. I recommend that we do that migration instead of the sqlite migration.

corneyl commented:
@MartinHjelmare I agree saving the key in a config entry is way nicer, and I can have a look at creating this, but:
- Requiring a config flow will vastly delay this bugfix PR
- Adding the config flow is quite out-of-scope for my feeling
- Until now, the integration also saves the key in a file to it's own liking, making for the same overhead

So I think this is more a choice between:
- A perfect integration with config flow, or
- An actually working, but less perfect, integration where the currently known bug is fixed

Do you think adding the config flow is worth waiting for?

MartinHjelmare commented:
That's my recommendation, yes.

balloob commented on homeassistant/components/webostv/__init__.py:
```diff
@@ -101,13 +104,43 @@ async def async_service_handler(service):
 
     host = conf[CONF_HOST]
     config_file = hass.config.path(WEBOSTV_CONFIG_FILE)
+    await asyncio.get_running_loop().run_in_executor(
+        None, convert_client_keys, config_file
```
```suggestion
    await hass.async_add_executor_job(
        convert_client_keys, config_file
```

balloob commented on tests/components/webostv/test_media_player.py:
```diff
@@ -52,6 +56,11 @@ async def setup_webostv(hass):
     await hass.async_block_till_done()
 
 
+def cleanup_webostv(hass):
+    """Test cleanup, remove the config file."""
+    os.remove(hass.config.path(WEBOSTV_CONFIG_FILE))
```
Make this a pytest fixture so it's also executed when the test fails.
```suggestion
@pytest.fixture
def cleanup_config(hass):
    """Test cleanup, remove the config file."""
    yield
    os.remove(hass.config.path(WEBOSTV_CONFIG_FILE))
```

balloob commented on tests/components/webostv/test_media_player.py:
```diff
@@ -128,3 +137,41 @@ async def test_command_with_optional_arg(hass, client):
     )
+
+
+async def test_migrate_keyfile_to_sqlite(hass, client):
+    """Test migration from JSON key-file to Sqlite based one."""
+    key = "3d5b1aeeb98e"
+    # Create config file with JSON content
+    config_file = hass.config.path(WEBOSTV_CONFIG_FILE)
+    with open(config_file, "w+") as file:
+        json.dump({"host": key}, file)
+
+    # Run the component setup
+    await setup_webostv(hass)
+
+    # Assert that the config file is a Sqlite database which contains the key
+    with SqliteDict(config_file) as conf:
+        assert conf.get("host") == key
+
+    cleanup_webostv(hass)
+
+
+async def test_dont_migrate_sqlite_keyfile(hass, client):
+    """Test that migration is not performed and setup still succeeds when config file is already an Sqlite DB."""
+    key = "3d5b1aeeb98e"
+
+    # Create config file with Sqlite DB
+    config_file = hass.config.path(WEBOSTV_CONFIG_FILE)
+    with SqliteDict(config_file) as conf:
+        conf["host"] = key
+        conf.commit()
+
+    # Run the component setup
+    await setup_webostv(hass)
+
+    # Assert that the config file is still an Sqlite database and setup didn't fail
+    with SqliteDict(config_file) as conf:
+        assert conf.get("host") == key
+
+    cleanup_webostv(hass)
```
```suggestion
async def test_migrate_keyfile_to_sqlite(hass, client, cleanup_config):
    """Test migration from JSON key-file to Sqlite based one."""
    key = "3d5b1aeeb98e"
    # Create config file with JSON content
    config_file = hass.config.path(WEBOSTV_CONFIG_FILE)
    with open(config_file, "w+") as file:
        json.dump({"host": key}, file)

    # Run the component setup
    await setup_webostv(hass)

    # Assert that the config file is a Sqlite database which contains the key
    with SqliteDict(config_file) as conf:
        assert conf.get("host") == key


async def test_dont_migrate_sqlite_keyfile(hass, client, cleanup_config):
    """Test that migration is not performed and setup still succeeds when config file is already an Sqlite DB."""
    key = "3d5b1aeeb98e"

    # Create config file with Sqlite DB
    config_file = hass.config.path(WEBOSTV_CONFIG_FILE)
    with SqliteDict(config_file) as conf:
        conf["host"] = key
        conf.commit()

    # Run the component setup
    await setup_webostv(hass)

    # Assert that the config file is still an Sqlite database and setup didn't fail
    with SqliteDict(config_file) as conf:
        assert conf.get("host") == key
```

balloob reviewed (approved):
Fine to merge as a fix. Migration can happen in a future PR.

corneyl committed:
Use right way of running sync jobs, improve tests using fixtures

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

corneyl committed:
Fix failing black check
